drivers/chibios/serial_usart_duplex.c

```diff
@@ -0,0 +1,245 @@
+#include "quantum.h"
+#include "serial.h"
+#include "printf.h"
+
+#include "ch.h"
+#include "hal.h"
+
+#ifndef USART_CR1_M0
+#    define USART_CR1_M0 USART_CR1_M  // some platforms (f1xx) dont have this so
+#endif
+
+#ifndef USE_GPIOV1
+// The default PAL alternate modes are used to signal that the pins are used for USART
+#    ifndef SERIAL_USART_TX_PAL_MODE
+#        define SERIAL_USART_TX_PAL_MODE 7
+#    endif
+#endif
+
+#ifndef SERIAL_USART_DRIVER
+#    define SERIAL_USART_DRIVER UARTD1
+#endif
+
+#ifndef SERIAL_USART_CR1
+#    define SERIAL_USART_CR1 (USART_CR1_PCE | USART_CR1_PS | USART_CR1_M0)  // parity enable, odd parity, 9 bit length
+#endif
+
+#ifndef SERIAL_USART_CR2
+#    define SERIAL_USART_CR2 0x0  // 1 stop bits
+#endif
+
+#ifndef SERIAL_USART_CR3
+#    define SERIAL_USART_CR3 0x0
+#endif
+
+#ifdef SERIAL_PIN_TX
+#    define SERIAL_USART_TX_PIN SERIAL_PIN_TX
+#endif
+
+#ifdef SERIAL_PIN_RX
+#    define SERIAL_USART_RX_PIN SERIAL_PIN_RX
+#endif
+
+#ifndef SELECT_SERIAL_SPEED
+#    define SELECT_SERIAL_SPEED 1
+#endif
+
+#ifdef SERIAL_USART_SPEED
+// Allow advanced users to directly set SERIAL_USART_SPEED
+#elif SERIAL_SPEED == 0
+#    define SERIAL_USART_SPEED 460800
+#elif SERIAL_SPEED == 1
+#    define SERIAL_USART_SPEED 230400
+#elif SERIAL_SPEED == 2
+#    define SERIAL_USART_SPEED 115200
+#elif SERIAL_SPEED == 3
+#    define SERIAL_USART_SPEED 57600
+#elif SERIAL_SPEED == 4
+#    define SERIAL_USART_SPEED 38400
+#elif SERIAL_SPEED == 5
+#    define SERIAL_USART_SPEED 19200
+#else
+#    error invalid SERIAL_SPEED value
+#endif
+
+#ifndef SERIAL_TIMEOUT_HANDSHAKE
+#    define SERIAL_TIMEOUT_HANDSHAKE 100
+#endif
+
+#ifndef SERIAL_TIMEOUT_BUFFER
+#    define SERIAL_TIMEOUT_BUFFER 100
+#endif
+
+#define HANDSHAKE_MAGIC 7
+
+/*
+ * UART driver configuration structure. Only the blocking DMA enabled API is used,
+ * therefore all callback functions are set to NULL.
+ */
+static UARTConfig uart_config = {NULL, NULL, NULL, NULL, NULL, (SERIAL_USART_SPEED), (SERIAL_USART_CR1), (SERIAL_USART_CR2), (SERIAL_USART_CR3)};
+
+static SSTD_t* Transaction_table      = NULL;
+static uint8_t Transaction_table_size = 0;
+
+void handle_transactions_target(void);
+
+__attribute__((weak)) void usart_init(void) {
+#if defined(USE_GPIOV1)
+    palSetLineMode(SERIAL_USART_TX_PIN, PAL_MODE_STM32_ALTERNATE_PUSHPULL);
+    palSetLineMode(SERIAL_USART_RX_PIN, PAL_MODE_INPUT);
+#else
+    palSetLineMode(SERIAL_USART_TX_PIN, PAL_MODE_ALTERNATE(SERIAL_USART_TX_PAL_MODE) | PAL_STM32_OTYPE_PUSHPULL | PAL_STM32_OSPEED_HIGHEST);
+    palSetLineMode(SERIAL_USART_RX_PIN, PAL_STM32_MODE_INPUT);
```

Comment:
This would also need to handle AF mode as per TX pin

Reply:
Could you elaborate on this? To my understanding the TX and RX pins are already correctly configured.

Reply:
Introduced SERIAL_USART_RX_PAL_MODE for recent mcus like the f303 only had experience with the f103 and see the point now.